File: lib/shared/ui/decision_request_markdown.dart
import 'dart:convert';

import '../util/decision_payload.dart';
import 'decision_copy.dart';

final _requestFence = RegExp(
  r'^```openhand-decision-request[ \t]*\r?\n([\s\S]*?)^```[ \t]*(?:\r?\n|$)',
  multiLine: true,
);
final _markdownPunctuation = RegExp(r'[\\`*_{}\[\]()<>#+.!|~\-]');

String _literal(String text) => text
    .replaceAllMapped(_markdownPunctuation, (match) => '\\${match[0]}')
    .replaceAll('\r\n', '\n')
    .replaceAll('\n', '  \n');

String _value(Object? value) => value is String
    ? _literal(value)
    : '```json\n${const JsonEncoder.withIndent('  ').convert(value).replaceAll('`', r'\u0060')}\n```';

/// 仅转换明确标记且有效的请求；原文由消息层保留，损坏内容不静默丢弃。
String? decisionRequestToMarkdown(String source, DecisionCopy copy) {
  if (source.length > DecisionPayload.maxCharacters ||
      !source.contains(DecisionPayload.requestLanguage)) {
    return null;
  }
  final matches = _requestFence.allMatches(source).toList();
  if (matches.isEmpty) return null;
  final output = StringBuffer();
  var end = 0;
  for (final match in matches) {
    if (!match[1]!.trimLeft().startsWith('{')) return null;
    final request = DecisionPayload.tryRequest(match[1]!);
    if (request == null) return null;
    output.writeln(_literal(source.substring(end, match.start)));
    output.writeln(
      '### ${copy.requestTitle}\n\n**${copy.stateLabel}**\n\n${_value(request['state'])}\n',
    );
    final questions = request['questions'] as Map;
    for (final entry in questions.entries) {
      final question = entry.value as Map;
      final type = question['type'] as String;
      output.writeln(
        '#### ${_literal(entry.key as String)} · ${copy.typeLabel(type)}\n\n${_value(question['instructions'])}\n',
      );
      final criteria = question['criteria'];
      if (criteria == null) continue;
      final label = type == DecisionPayload.typeScore
          ? copy.scoreItemLabel
          : type == DecisionPayload.typeChoice
          ? copy.choiceItemLabel
          : copy.criteriaLabel;
      output.writeln('**$label**\n');
      final items = criteria is Map
          ? criteria.entries.toList()
          : (criteria as List).asMap().entries.toList();
      for (var index = 0; index < items.length; index++) {
        final entry = items[index];
        final value = criteria is Map
            ? '${_literal('${entry.key}')}${entry.value == null
                  ? ''
                  : entry.value is String
                  ? '：${_value(entry.value)}'
                  : '\n\n${_value(entry.value)}'}'
            : _value(entry.value);
        output.writeln('${index + 1}. ${value.replaceAll('\n', '\n   ')}');
      }
      output.writeln();
    }
    end = match.end;
  }
  output.write(_literal(source.substring(end)));
  return output.toString().trim();
}

File: clients/web/src/shared/util/decision_request_markdown.ts
import { t } from '../../i18n';
import { DECISION_MAX_CHARACTERS, DECISION_REQUEST, parseDecisionRequest } from './decision';

const literal = (text: string): string => text.replace(/[\\`*_{}\[\]()<>#+.!|~\-]/g, '\\$&').replace(/\r\n/g, '\n').replace(/\n/g, '  \n');
const valueText = (value: unknown): string => typeof value === 'string' ? literal(value)
  : '```json\n' + JSON.stringify(value, null, 2).replace(/`/g, '\\u0060') + '\n```';

/** 展示转换不修改消息原文；任一请求损坏时整体回退，避免遗漏输入。 */
export function decisionRequestToMarkdown(source: string): string | null {
  if (source.length > DECISION_MAX_CHARACTERS || !source.includes(DECISION_REQUEST)) return null;
  const matches = [...source.matchAll(/^```openhand-decision-request[ \t]*\r?\n([\s\S]*?)^```[ \t]*(?:\r?\n|$)/gm)];
  if (!matches.length) return null;
  const output: string[] = [];
  let end = 0;
  for (const match of matches) {
    const request = parseDecisionRequest(match[1]);
    if (!request) return null;
    output.push(literal(source.slice(end, match.index)));
    output.push(`### ${t('decision.request.title', '结构化决策')}\n\n**${t('decision.field.state', '待评估内容')}**\n\n${valueText(request.state)}`);
    for (const [name, question] of Object.entries(request.questions)) {
      const type = question.type;
      output.push(`#### ${literal(name)} · ${t(`decision.type.${type}`, type)}\n\n${valueText(question.instructions)}`);
      const criteria = question.criteria;
      if (criteria == null) continue;
      const label = type === 'score' ? t('decision.field.scoreLevels', '评分等级（从低到高）')
        : type === 'choice' ? t('decision.field.options', '候选项') : t('decision.field.criteria', '判断标准');
      const items = Object.entries(criteria);
      output.push(`**${label}**\n\n` + items.map(([key, value], index) => {
        const text = Array.isArray(criteria) ? valueText(value) : `${literal(key)}${value == null ? '' : typeof value === 'string' ? `：${valueText(value)}` : `\n\n${valueText(value)}`}`;
        return `${index + 1}. ${text.replace(/\n/g, '\n   ')}`;
      }).join('\n'));
    }
    end = match.index! + match[0].length;
  }
  output.push(literal(source.slice(end)));
  return output.join('\n\n').trim();
}
